### PR TITLE
Add support for bucketized req size for attention metadata

### PR DESCRIPTION
**Pull request by helloworld1**
# Description

Many attention preprocessing are done on max_req_len. But if the actual req_len is much less than max_seq_lens, the padding added is a big waste of resource to compute them. 

In this PR, we will slice the q_start_loc and seq_len to the closest power of two of the actual req lens (padded_req_lens). This will greatly reduce the preprocessing time in attention

If the change fixes a Github issue, please include a link, e.g.,:
FIXES: #123456

# Tests

mmlu_pro evaluated, request rate 16, 64, 512 on bucket 64, max_req_size=512

Below shows req rate 64
```
|       Tasks       |Version|    Filter    |n-shot|  Metric   |   |Value |   |Stderr|
|-------------------|------:|--------------|-----:|-----------|---|-----:|---|-----:|
|mmlu_pro           |    2.0|custom-extract|     5|exact_match|↑  |0.8329|±  |0.0139|
| - biology         |    3.1|custom-extract|     5|exact_match|↑  |0.9400|±  |0.0339|
| - business        |    3.1|custom-extract|     5|exact_match|↑  |0.8600|±  |0.0496|
| - chemistry       |    3.1|custom-extract|     5|exact_match|↑  |0.9000|±  |0.0429|
| - computer_science|    3.1|custom-extract|     5|exact_match|↑  |0.8200|±  |0.0549|
| - economics       |    3.1|custom-extract|     5|exact_match|↑  |0.9200|±  |0.0388|
| - engineering     |    3.1|custom-extract|     5|exact_match|↑  |0.6800|±  |0.0666|
| - health          |    3.1|custom-extract|     5|exact_match|↑  |0.7800|±  |0.0592|
| - history         |    3.1|custom-extract|     5|exact_match|↑  |0.6800|±  |0.0666|
| - law             |    3.1|custom-extract|     5|exact_match|↑  |0.8000|±  |0.0571|
| - math            |    3.1|custom-extract|     5|exact_match|↑  |0.9600|±  |0.0280|
| - other           |    3.1|custom-extract|     5|exact_match|↑  |0.7800|±  |0.0592|
| - philosophy      |    3.1|custom-extract|     5|exact_match|↑  |0.8400|±  |0.0524|
| - physics         |    3.1|custom-extract|     5|exact_match|↑  |0.9200|±  |0.0388|
| - psychology      |    3.1|custom-extract|     5|exact_match|↑  |0.7800|±  |0.0592|

| Groups |Version|    Filter    |n-shot|  Metric   |   |Value |   |Stderr|
|--------|------:|--------------|-----:|-----------|---|-----:|---|-----:|
|mmlu_pro|      2|custom-extract|     5|exact_match|↑  |0.8329|±  |0.0139|
```

8K/1K Prefill, max_concurrency=64, max_req_size=512

Before:
conv1d: 301 us
gdn schedule: 283 us
gdn kernel: 834 us
gdn+conv: 1418us

After (With bucket [64,512]):
conv1d: 126us
gdn schedule: 97us
gdn kernel: 855 us
gdn+conv: 1078 us

E2E:
Before: 2328 total tokens / chip / s 
After: 2516 total tokens / chip / s

# Checklist

Before submitting this PR, please make sure:
- I have performed a self-review of my code.
- I have necessary comments in my code, particularly in hard-to-understand areas.
- I have made or will make corresponding changes to any relevant documentation.

### CODE REVIEW

File: tpu_inference/layers/common/attention_metadata.py [hunk 1]
```diff
@@ -29,7 +29,7 @@
         "request_distribution",
         "mamba_state_indices",
     ],
-    meta_fields=[],
+    meta_fields=["padded_num_reqs"],
```
Comment (jrplatin) [inline tpu_inference/layers/common/attention_metadata.py:32]: Why are you considering this a meta field and not a data field?

File: tpu_inference/envs.py [hunk 2]
```diff
@@ -234,6 +255,16 @@ def _get_str_list_env() -> list[str]:
     "MOE_REQUANTIZE_BLOCK_SIZE":
     lambda: int(block_size)
     if (block_size := os.getenv("MOE_REQUANTIZE_BLOCK_SIZE")) else None,
+    # Bucketize the num reqs instead of max reqs for attention to improve
```
Comment (jrplatin) [inline tpu_inference/envs.py]: I would expand on what this means -- what do you mean by "bucketize" ?

File: tpu_inference/runner/utils.py [hunk 3]
```diff
@@ -68,6 +68,26 @@ def get_req_paddings(min_req_size: int, max_req_size: int) -> list[int]:
...
+    """Get num reqs paddings with custom override to reduce compilation time"""
+    if not envs.ATTN_BUCKETIZED_NUM_REQS:
+        reqs = [max_req_size]
+    elif envs.ATTN_CUSTOM_NUM_REQS_BUCKETS:
+        reqs = envs.ATTN_CUSTOM_NUM_REQS_BUCKETS
+    else:
+        reqs = get_req_paddings(min_req_size, max_req_size)
+
+    if max_req_size not in reqs:
+        logger.info(
```
Comment (jrplatin) [inline tpu_inference/runner/utils.py:84]: When would this be the case?

File: tpu_inference/runner/tpu_runner.py [hunk 4]
```diff
@@ -524,6 +524,11 @@ def _init_inputs(self) -> None:
         min_num_reqs = max(MIN_NUM_SEQS, next_power_of_2(self.dp_size))
         self.num_reqs_paddings = runner_utils.get_req_paddings(
             min_req_size=min_num_reqs, max_req_size=self.max_num_reqs)
+
+        # The num_reqs paddings for attention only since the padding can
+        # have custom overrides
```
Comment (jrplatin) [inline tpu_inference/runner/tpu_runner.py]: Might be good to expand on this and explain the bucketing and relevant env vars and such

File: tpu_inference/layers/common/attention_metadata.py (hunk 1 above)
Comment (helloworld1) [reply tpu_inference/layers/common/attention_metadata.py:32]: meta_fields are static values (int instead of jnp.int32). It allows statically slicing arrays, which is much more performant and simpler in jit.compile than dynamic slicing.

File: tpu_inference/runner/tpu_runner.py (hunk 4 above)
Comment (helloworld1) [reply tpu_inference/runner/tpu_runner.py]: Added comments

File: tpu_inference/runner/utils.py (hunk 3 above)
Comment (helloworld1) [reply tpu_inference/runner/utils.py:84]: User can set ATTN_CUSTOM_NUM_REQS_BUCKETS that does not include the max_req_size. It is OK, and we just manually add max_req_size. It is telling user max_reqs is added in additional to what user set.

File: tpu_inference/envs.py (hunk 2 above)
Comment (helloworld1) [reply tpu_inference/envs.py]: Rewrite the comments to make it much more clear

**Review by jrplatin — APPROVED** [review]
